Projects/Server/Items/Container.Enumerable.cs

```diff
@@ -0,0 +1,226 @@
+
+    public readonly ref struct AllItemsEnumerable<TFilter, TResult>
+        where TFilter : struct, IFilter<Item, TResult>
+    {
+        private readonly Container container;
+        private readonly ref TFilter filter;
+
+        public AllItemsEnumerable(Container container, scoped in TFilter filter = default)
+        {
+            this.container = container;
+            this.filter = filter;
+        }
+
+        public AllItemsEnumerable<ChainFilter<Item, TResult, TTResult, TFilter, TTFilter>, TTResult> Where<TTFilter, TTResult>(TTFilter filter = default)
+            where TTFilter : struct, IFilter<TResult, TTResult>
+        {
+            return new(container, new(this.filter, filter));
+        }
+
+        public AllItemsEnumerable<ChainFilter<Item, TResult, TTResult, TFilter, TypeFilter<TResult, TTResult>>, TTResult> OfType<TTResult>()
+        {
+            return new(container, new(filter, default));
+        }
+
+        public AllItemsEnumerable<ChainFilter<Item, TResult, TResult, TFilter, TakeFilter<TResult>>, TResult> Take(int amount)
+        {
+            return new(container, new(filter, new(amount)));
+        }
+
+        public TResult? FirstOrDefault(TResult? @default = default)
+        {
+            if (container.m_Items is not { Count: > 0 })
+            {
+                return @default;
+            }
+
+            Container? current = container;
+
+            do
+            {
+                Span<Item> itemsSpan = CollectionsMarshal.AsSpan(current.m_Items);
+
+                for (int i = 0; i < itemsSpan.Length; i++)
+                {
+                    Item item = itemsSpan[i];
+
+                    if (item is TResult tItem)
+                    {
+                        containersForGetItem.Clear();
+                        return tItem;
+                    }
+
+                    if (item is Container { m_Items.Count: > 0 } c)
+                    {
+                        containersForGetItem.Enqueue(c);
+                    }
+                }
+            }
+            while (containersForGetItem.TryDequeue(out current));
+
+            containersForGetItem.Clear();
+
+            return @default;
+        }
+
+        public bool Any() => FirstOrDefault() != null;
+
+        public PooledRefQueue<TResult> ToPooledRefQueue()
+        {
+            PooledRefQueue<TResult> toRet = new();
+
+            foreach (TResult item in this)
+            {
+                toRet.Enqueue(item);
+            }
+
+            return toRet;
+        }
+
+        public List<TResult> ToList()
+        {
+            List<TResult> toRet = new();
+
+            foreach (TResult item in this)
+            {
+                toRet.Add(item);
+            }
+
+            return toRet;
+        }
+
+        public TestAllItemsEnumerator<TFilter, TResult> GetEnumerator() => new(container, in filter);
+    }
+
+    public ref struct TestAllItemsEnumerator<TFilter, TResult>
```

Comment:
Technically Items can have `m_Items`, so this can be put generalized even more.
With that said, we can put this in it's own file I think?

Reply:
Yes there are similarities between Item and Container.
I can switch to "Items" property instead of m_Items, that is the only common accessor for both classes, so the code can be used with little modifications also for standard Item class if required.
I don't know the use cases of m_Items for standard Item class btw.